### PR TITLE
Feature/sprint 18

### PR DESCRIPTION
**Pull request by mikeymonster**

### CODE REVIEW

File: src/Sfa.Tl.Matching.Application/Extensions/TimeSpanExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using Humanizer;
+using Humanizer.Localisation;
+
+namespace Sfa.Tl.Matching.Application.Extensions
+{
+    public static class TimeSpanExtensions
+    {
+        public static string ToReadableString(this long? seconds, int precision = 2, TimeUnit minUnit = TimeUnit.Minute, string separator = " ")
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Application/Extensions/TimeSpanExtensions.cs]: this looks more like DateTimeProvider Method

File: src/Sfa.Tl.Matching.Application/Services/EmailService.cs [hunk 2]
```diff
@@ -13,32 +13,31 @@ namespace Sfa.Tl.Matching.Application.Services
...
             _configuration = configuration;
             _emailTemplateRepository = emailTemplateRepository;
-            _notificationsApi = notificationsApi;
+            _notificationClient = notificationClient;
             _logger = logger;
+
         }
 
-        public async Task SendEmail(string templateName, string toAddress, string subject, IDictionary<string, string> personalisationTokens, string replyToAddress)
+        public async Task SendEmailAsync(string templateName, string toAddress, IDictionary<string, string> personalisationTokens)
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Application/Services/EmailService.cs]: Why subject is removed?

File: src/Sfa.Tl.Matching.Application/Services/EmployerService.cs [hunk 3]
```diff
@@ -223,35 +229,72 @@ private async Task<int> CreateOrUpdateEmployerAsync(CrmEmployerEventBase employe
...
+            await _employerRepository.UpdateAsync(existingEmployer);
+
             return 1;
+        }
 
+        private async Task AddMessageToQueueAsync(CrmEmployerEventBase employerData)
+        {
+            await _messageQueueService.PushEmployerAupaBlankEmailMessageAsync(new SendEmployerAupaBlankEmail
+            {
+                Name = employerData.Name,
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Application/Services/EmployerService.cs]: can you please use Mapper here

File: src/Sfa.Tl.Matching.Application/Services/ServiceFactory/FeedbackFactory.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿using Sfa.Tl.Matching.Application.Interfaces;
+using Sfa.Tl.Matching.Application.Interfaces.ServiceFactory;
+
+namespace Sfa.Tl.Matching.Application.Services.ServiceFactory
+{
+    public class FeedbackFactory<T> : IFeedbackFactory<T>
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Application/Services/ServiceFactory/FeedbackFactory.cs]: This factory is not doing anything remove it

File: src/Sfa.Tl.Matching.Models/Dto/OpportunityItemWithUsernameForEmployerFeedbackSentDto.cs [hunk 5]
```diff
@@ -1,14 +0,0 @@
-﻿namespace Sfa.Tl.Matching.Models.Dto
-{
-    public class OpportunityItemWithUsernameForEmployerFeedbackSentDto
-    {
-        public int Id { get; set; }
-        public string Username { get; set; }
-    }
-
-    public class OpportunityItemWithUsernameForProviderFeedbackSentDto
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Models/Dto/OpportunityItemWithUsernameForEmployerFeedbackSentDto.cs]: put this in separate file

File: src/Sfa.Tl.Matching.Web/Controllers/OpportunityController.cs [hunk 6]
```diff
@@ -66,9 +66,10 @@ public async Task<IActionResult> SaveProvisionGap(SaveProvisionGapViewModel view
         }
 
         [Route("referral-create", Name = "SaveReferral")]
-        public async Task<IActionResult> SaveReferral(string viewModel)
+        public async Task<IActionResult> SaveReferralAsync()
         {
-            var saveReferralViewModel = JsonConvert.DeserializeObject<SaveReferralViewModel>(viewModel);
+            var selectedProviders = TempData["SelectedProviders"] as string;
```
Comment (hardik-desai-kcl) [inline src/Sfa.Tl.Matching.Web/Controllers/OpportunityController.cs]: why temp data? can use hidden field something else? db cache some thing else

File: src/Sfa.Tl.Matching.Api.Clients.UnitTests/Factories/GoogleMapsHttpClientFactory.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
+using System.Net.Http;
+using Sfa.Tl.Matching.Api.Clients.GoogleMaps;
+
+namespace Sfa.Tl.Matching.Api.Clients.UnitTests.Factories
```
[the remark this answers is not in the document]
Comment (mikeymonster) [reply src/Sfa.Tl.Matching.Api.Clients.UnitTests/Factories/GoogleMapsHttpClientFactory.cs]: Fixed on new sprint branch

File: src/Sfa.Tl.Matching.Api.Clients/GoogleDistanceMatrix/GoogleDistanceMatrixApiClient.cs [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+                uriBuilder.Append("units=imperial");
+                //uriBuilder.Append($"&origins={latitude}%2C{longitude}");
+                uriBuilder.Append($"&origins={WebUtility.UrlEncode(origin)}");
+
+                uriBuilder.Append($"&mode={travelMode}");
+                uriBuilder.Append($"&arrival_time={arrivalTimeSeconds}");
+                uriBuilder.Append("&destinations=");
+
+                //uriBuilder.Append("enc:");
+                //var polyline = EncodePolyline(destinations);
```
[the remark this answers is not in the document]
Comment (mikeymonster) [reply src/Sfa.Tl.Matching.Api.Clients/GoogleDistanceMatrix/GoogleDistanceMatrixApiClient.cs]: Done in new sprint branch

File: src/Sfa.Tl.Matching.Api.Clients/GoogleDistanceMatrix/GoogleDistanceMatrixApiClient.cs [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError($"Failure calling google api - {ex}");
+                throw;
+            }
+        }
+
+        //Implementation from https://gist.github.com/shinyzhu/4617989
+        public string EncodePolyline(IList<LocationDto> points)
```
[the remark this answers is not in the document]
Comment (mikeymonster) [reply src/Sfa.Tl.Matching.Api.Clients/GoogleDistanceMatrix/GoogleDistanceMatrixApiClient.cs]: Removed this method in new sprint branch

File: src/Sfa.Tl.Matching.Api.Clients/GeoLocations/ILocationApiClient.cs [hunk 10]
```diff
@@ -5,7 +5,11 @@ namespace Sfa.Tl.Matching.Api.Clients.GeoLocations
 {
     public interface ILocationApiClient
     {
-        Task<(bool, string)> IsValidPostCode(string postCode);
-        Task<PostCodeLookupResultDto> GetGeoLocationData(string postCode);
+        Task<(bool, string)> IsValidPostcodeAsync(string postcode);
+        Task<(bool, string)> IsValidPostcodeAsync(string postcode, bool includeTerminated);
+        Task<(bool, string)> IsTerminatedPostcodeAsync(string postcode);
+        Task<PostcodeLookupResultDto> GetGeoLocationDataAsync(string postcode);
+        Task<PostcodeLookupResultDto> GetGeoLocationDataAsync(string postcode, bool includeTerminated);
```
[the remark this answers is not in the document]
Comment (mikeymonster) [reply src/Sfa.Tl.Matching.Api.Clients/GeoLocations/ILocationApiClient.cs]: Refactored to have a single method with explicit parameters.

File: src/Sfa.Tl.Matching.Api.Clients/GoogleDistanceMatrix/GoogleDistanceMatrixApiClient.cs [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+            _httpClient.DefaultRequestHeaders.Accept.Clear();
+            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        }
+
+        public async Task<IDictionary<int, JourneyInfoDto>> GetJourneyTimesAsync(string originPostcode, decimal latitude, decimal longitude,
+            IList<LocationDto> destinations, string travelMode, long arrivalTimeSeconds)
+        {
+            const int batchSize = 100; //Max client-side elements: 100
+            var batches = CreateBatches(destinations, batchSize);
```
[the remark this answers is not in the document]
Comment (mikeymonster) [reply src/Sfa.Tl.Matching.Api.Clients/GoogleDistanceMatrix/GoogleDistanceMatrixApiClient.cs]: Changed to use a list in new sprint branch.

File: src/Sfa.Tl.Matching.Api.Clients/GoogleDistanceMatrix/GoogleDistanceMatrixApiClient.cs [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+            var distanceSearchResults = new Dictionary<int, JourneyInfoDto>(batches.Count);
+            
+            foreach (var batch in batches)
+            {
+                var (_, value) = batch;
+                var response = SearchBatchAsync(originPostcode, latitude, longitude, value, travelMode, arrivalTimeSeconds).GetAwaiter().GetResult();
```
[the remark this answers is not in the document]
Comment (mikeymonster) [reply src/Sfa.Tl.Matching.Api.Clients/GoogleDistanceMatrix/GoogleDistanceMatrixApiClient.cs]: Changed the call to use await - the method is definitely async but the call used to be inside a parallel foreach. Changes will be done on new sprint branch.

**Pull request merged**